File: src/main/java/net/tywrapstudios/blossombridge/api/logging/LoggingHandler.java
package net.tywrapstudios.blossombridge.api.logging;

import net.tywrapstudios.blossombridge.api.config.AbstractConfig;
import net.tywrapstudios.blossombridge.api.config.ConfigManager;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * A general-purpose Logging Handler compatible with the modular configuration system.
 * This class dynamically fetches configuration values for logging behavior.
 * <p> Note that this should only be used when a Main Config Object (which have to extend {@link AbstractConfig} anyway) object is available.
 * @param <T> The Main Config class of the mod, which is usually already an instance of {@link AbstractConfig}.
 * @author Tiazzz
 */
public class LoggingHandler<T extends AbstractConfig> {
    private final Logger MAIN;
    private final Logger DEBUG;
    private final ConfigManager<T> manager;

    /**
     * Construct a new {@link LoggingHandler}.
     * @param mainLogger A {@link Logger} that logs the main parts of the mod.
     * @param debugLogger A {@link Logger} that logs the debug parts of the mod.
     * @param configManager Your {@link ConfigManager}, this is needed to get various values from your Config.
     */
    public LoggingHandler(Logger mainLogger, Logger debugLogger, ConfigManager<T> configManager) {
        this.MAIN = mainLogger;
        this.DEBUG = debugLogger;
        this.manager = configManager;
    }

    /**
     * Construct a new {@link LoggingHandler}.
     * @param name A {@link String} that represents your mod's name, {@link Logger}{@code s} will automatically be made for you.
     * @param configManager Your {@link ConfigManager}, this is needed to get various values from your Config.
     */
    public LoggingHandler(String name, ConfigManager<T> configManager) {
        this.MAIN = LoggerFactory.getLogger(name + "|Main");
        this.DEBUG = LoggerFactory.getLogger(name + "|Debug");
        this.manager = configManager;
    }

    /**
     * Log a String to the info level of the main logger.
     *
     * @param message The Message String to log.
     */
    public void info(String message) {
        MAIN.info(message);
    }

    /**
     * Log a String to the warn level of the main logger.
     *
     * @param message The Message String to log.
     */
    public void warn(String message) {
        if (!conf().util_config.suppress_warns) {
            MAIN.warn(message);
        }
    }

    /**
     * Log a String to the error level of the main logger.
     *
     * @param message The Message String to log.
     */
    public void error(String message) {
        if (!conf().util_config.suppress_warns) {
            MAIN.error(message);
        }
    }

    /**
     * Log a String to the info and debug level of the debug logger.
     *
     * @param message The Message String to log.
     */
    public void debug(String message) {
        if (conf().util_config.debug_mode) {
            DEBUG.info(message);
        }
        literalDebug(message);
    }

    /**
     * Log a String to the debug level of the debug logger.
     *
     * @param message The Message String to log.
     */
    public void literalDebug(String message) {
        if (conf().util_config.debug_mode) {
            DEBUG.debug(message);
        }
    }

    /**
     * Alternative version of {@link #literalDebug(String)} that optionally ignores config settings.
     *
     * @param message     The Message String to log.
     * @param ignoreConfig Whether to ignore any Config settings.
     */
    public void literalDebug(String message, boolean ignoreConfig) {
        if (ignoreConfig) {
            DEBUG.debug(message);
        } else {
            literalDebug(message);
        }
    }

    /**
     * Log a String to the warn level of the debug logger.
     * <p>Note that this will also send a literalDebug, which is on the debug level, but it simply has {@code [WARN]} prefixed to it.
     *
     * @param message The Message String to log.
     */
    public void debugWarning(String message) {
        if (conf().util_config.debug_mode && !conf().util_config.suppress_warns) {
            DEBUG.warn(message);
        }
        literalDebug("[WARN] " + message);
    }

    private AbstractConfig conf() {
        return manager.getConfig();
    }
}

File: src/main/kotlin/org/tywrapstudios/blossombridge/api/logging/LoggingHandler.kt
package org.tywrapstudios.blossombridge.api.logging

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.tywrapstudios.blossombridge.api.config.BasicConfigClass
import org.tywrapstudios.blossombridge.api.config.ConfigManager

/**
 * A general-purpose Logging Handler compatible with the modular configuration system.
 * This class dynamically fetches configuration values for logging behavior.
 *
 * Note that this should only be used when a Main Config Object (which have to extend [BasicConfigClass] anyway) object is available.
 */
class LoggingHandler<T : BasicConfigClass> {
    private val main: Logger
    private val debug: Logger
    private val manager: ConfigManager<T>

    /**
     * Construct a new [LoggingHandler].
     * @param mainLogger A [Logger] that logs the main parts of the mod.
     * @param debugLogger A [Logger] that logs the debug parts of the mod.
     * @param configManager Your [ConfigManager], this is needed to get various values from your Config.
     */
    constructor(mainLogger: Logger, debugLogger: Logger, configManager: ConfigManager<T>) {
        this.main = mainLogger
        this.debug = debugLogger
        this.manager = configManager
    }

    /**
     * Construct a new [LoggingHandler].
     * @param name A [String] that represents your mod's name, [Logger]`s` will automatically be made for you.
     * @param configManager Your [ConfigManager], this is needed to get various values from your Config.
     */
    constructor(name: String, configManager: ConfigManager<T>) {
        this.main = LoggerFactory.getLogger("$name|Main")
        this.debug = LoggerFactory.getLogger("$name|Debug")
        this.manager = configManager
    }

    /**
     * Log a String to the info level of the main logger.
     *
     * @param message The Message String to log.
     */
    fun info(message: String) {
        main.info(message)
    }

    /**
     * Log a String to the warn level of the main logger.
     *
     * @param message The Message String to log.
     */
    fun warn(message: String) {
        if (!conf().util_config.suppress_warns) {
            main.warn(message)
        }
    }

    /**
     * Log a String to the error level of the main logger.
     *
     * @param message The Message String to log.
     */
    fun error(message: String) {
        if (!conf().util_config.suppress_warns) {
            main.error(message)
        }
    }

    /**
     * Log a String to the info and debug level of the debug logger.
     *
     * @param message The Message String to log.
     */
    fun debug(message: String) {
        if (conf().util_config.debug_mode) {
            debug.info(message)
        }
        literalDebug(message)
    }

    /**
     * Log a String to the debug level of the debug logger.
     *
     * @param message The Message String to log.
     */
    fun literalDebug(message: String) {
        debug.debug(message)
    }

    /**
     * Log a String to the warn level of the debug logger.
     *
     * @param message The Message String to log.
     */
    fun debugWarning(message: String) {
        if (conf().util_config.debug_mode && !conf().util_config.suppress_warns) {
            debug.warn(message)
        }
    }

    private fun conf(): T {
        return manager.getConfig()!!
    }
}